### PR TITLE
GEOS-11461: Support MapML embedded viewer on HTML for WFS GetFeature

### PR DESCRIPTION
**Pull request by dromagnoli**
[GEOS-11461](https://osgeo-org.atlassian.net/browse/GEOS-11461) [Powered by Pull Request Badge](https://pullrequestbadge.com/?utm_medium=github&utm_source=geoserver&utm_campaign=badge_info)

  

[GEOS-11461](https://osgeo-org.atlassian.net/browse/GEOS-11461)

# Checklist

- [x] I have read the [contribution guidelines](https://github.com/geoserver/geoserver/blob/main/CONTRIBUTING.md).
- [x] I have sent a [Contribution Licence Agreement](https://docs.geoserver.org/latest/en/developer/policies/committing.html) (not required for small changes, e.g., fixing typos in documentation).
- [x] First PR targets the `main` branch (backports managed later; ignore for branch specific issues).
- [x] All the build checks are green ([see automated QA checks](https://docs.geoserver.org/latest/en/developer/qa-guide/index.html)).

For core and extension modules:

- [x] New unit tests have been added covering the changes.
- [x] [Documentation](https://github.com/geoserver/geoserver/tree/main/doc/en/user/source) has been updated (if change is visible to end users).
- [ ] The [REST API docs](https://github.com/geoserver/geoserver/tree/main/doc/en/api/1.0.0) have been updated (when changing configuration objects or the REST controllers).
- [x] There is an issue in the [GeoServer Jira](https://osgeo-org.atlassian.net/browse/GEOS/summary) (except for changes that do not affect administrators or end users in any way).
- [x] Commit message(s) must be in the form ``[GEOS-XYZWV] Title of the Jira ticket``.
- [x] Bug fixes and small new features are presented as a single commit.
- [x] Each commit has a single objective (if there are multiple commits, each has a separate JIRA ticket describing its goal).

### CODE REVIEW

**Review by aaime — CHANGES_REQUESTED** [review]
Overall looks good. I was surprised a bit when trying to open a "coastlines" layer we have in the default data director, as it provided a partial map:

image


Then I realized it was the default limit of 50 features. However, removing it, led the browser to "brick" itself, too much data, it became non responsive:

image

and had to kill it (using Chrome internal tasks manager).
Using Firefox results in the same.

So... maybe add a warning in the documentation, that removing the max feature limit for the mapml WFS output format can lead to browser issues?

I've also tried it with spearfish roads, which is pretty small, and it had the same effect... ouch? I get that the browser will be eventually unable to handle all these vectors, but getting stuck on that layer seems too much?

@prushforth is this to be expected? The SF:roads dumped into mapml (WFS style) is only 500KB.

File: src/extension/mapml/src/main/java/org/geoserver/mapml/MapMLHTMLOutput.java [hunk 1]
```diff
@@ -0,0 +1,176 @@
...
+                CRS.getAxisOrder(projectedBbox.getCoordinateReferenceSystem())
+                        .equals(CRS.AxisOrder.NORTH_EAST);
+        double minX = flipAxis ? projectedBbox.getMinY() : projectedBbox.getMinX();
+        double maxX = flipAxis ? projectedBbox.getMaxY() : projectedBbox.getMaxX();
+        double minY = flipAxis ? projectedBbox.getMinX() : projectedBbox.getMinY();
+        double maxY = flipAxis ? projectedBbox.getMaxX() : projectedBbox.getMaxY();
+        final Bounds pb = new Bounds(new Point(minX, minY), new Point(maxX, maxY));
+        // allowing for the data to be displayed at 1024x768 pixels, figure out
+        // the zoom level at which the projected bounds fits into 1024x768
+        // in both dimensions
```
[the remark this answers is not in the document]
Comment (dromagnoli) [reply src/extension/mapml/src/main/java/org/geoserver/mapml/MapMLHTMLOutput.java:170]: That part comes from an existing method that has been moved. I'm going to change that comment

**Comment by prushforth** [comment]
> @prushforth is this to be expected? The SF:roads dumped into mapml (WFS style) is only 500KB.

It is, a little bit, yes: loading all the features from a dataset into a map view at small scale is a bit of what WFS brings you by default, with no opportunity on the server to cull, generalize etc because no scale information is passed. The current approach of linking via \<layer- src="URL to text/mapml for layer"> is loading effectively all the features of the layer into the map at once.  We would 'normally' use a bbox in a web map to constrain how many features are loaded into a map view, but of course the bbox at a small scale would include all features in the layer so it's a conundrum.  

By using a different approach in \<layer-> in our text/html; subtype=mapml, we could possibly improve the performance: by using an inline \<map-extent> to filter the features by bbox would (potentially) limit what gets returned; set the initial view of the viewer to some scale large enough to manage the load, limit the zoom range of the \<map-extent> to something less than say 200 features (how many features is useful / too many from a users POV?). The client used to support feature paging  but it complicated the client and didn't improve the user experience so we dropped it.

**Comment by aaime** [comment]
Ok, I don't see much that can be reasonably be done out of the box then... trying to zoom in a specific bbox requires some heuristic to find such bbox, and it could be pretty expensive to implement, especially on datasets that lack uniform data distribution. For example, OSM building layers... it's way too massive to be displayed at a global level, and yet, finding a specific bbox that has say 200 features would take some lookup time... and navigation would be broken, as users would naturally tend to zoom out to find another city, bogging down the client as a result.

Having a 50 features limit seems to work out in this case.

**Comment by prushforth** [comment]
Agreed.

**Review by aaime — APPROVED** [review]

**Pull request merged**